src/read.rs

```diff
@@ -549,264 +597,20 @@ impl<R: Read + Seek> ZipArchive<R> {
-                    &mut invalid_errors_64,
-                    &mut unsupported_errors_64,
-                    &mut inner_results,
-                    &(),
-                );
-                inner_results.into_iter().for_each(|(_, results)| {
-                    results.into_iter().for_each(|result| {
-                        Self::sort_result(
-                            result,
-                            &mut invalid_errors_64,
-                            &mut unsupported_errors_64,
-                            &mut ok_results,
-                            &footer,
-                        );
-                    });
-                });
-            });
-        ok_results.sort_by_key(|(_, result)| {
-            (
-                u64::MAX - result.cde_position, // try the last one first
-                !result.is_zip64,               // try ZIP64 first
-            )
-        });
-        let mut best_result = None;
-        for (footer, result) in ok_results {
-            let mut inner_result = Vec::with_capacity(1);
-            let is_zip64 = result.is_zip64;
-            Self::sort_result(
-                Self::read_central_header(result, config, reader),
-                if is_zip64 {
-                    &mut invalid_errors_64
-                } else {
-                    &mut invalid_errors_32
-                },
-                if is_zip64 {
-                    &mut unsupported_errors_64
-                } else {
-                    &mut unsupported_errors_32
-                },
-                &mut inner_result,
-                &(),
-            );
-            if let Some((_, shared)) = inner_result.into_iter().next() {
-                if shared.files.len() == footer.number_of_files as usize
-                    || (is_zip64 && footer.number_of_files == ZIP64_ENTRY_THR as u16)
-                {
-                    best_result = Some((footer, shared));
-                    break;
-                } else {
-                    if is_zip64 {
-                        &mut invalid_errors_64
-                    } else {
-                        &mut invalid_errors_32
-                    }
-                    .push(InvalidArchive("wrong number of files"))
-                }
-            }
-        }
-        let Some((footer, shared)) = best_result else {
-            return Err(unsupported_errors_32
-                .into_iter()
-                .chain(unsupported_errors_64)
-                .chain(invalid_errors_32)
-                .chain(invalid_errors_64)
-                .next()
-                .unwrap());
-        };
-        reader.seek(SeekFrom::Start(shared.dir_start))?;
-        Ok((Rc::try_unwrap(footer).unwrap(), shared.build()))
+    pub(crate) fn get_metadata(config: Config, reader: &mut R) -> ZipResult<Shared> {
+        // Find the EOCD and possibly EOCD64 entries and determine the archive offset.
+        let cde = spec::find_central_directory(reader, config.archive_offset)?;
```

Comment:
What happens if something that looks like a valid EOCD or EOCD64 block, but doesn't have a valid central directory in front of it and thus fails `try_from`, is included in the file comment of a valid ZIP file? We should keep looking for the real one in that case.

Reply:
I'll wrap this in a loop and allow the `find_central_directory` function to continue from the previous EOCD candidate backwards.